freqtrade/strategy/interface.py

```diff
@@ -634,12 +634,49 @@ def min_roi_reached_entry(self, trade_dur: int) -> Tuple[Optional[int], Optional
         :param trade_dur: trade duration in minutes
         :return: minimal ROI entry value or None if none proper ROI entry was found.
         """
-        # Get highest entry in ROI dict where key <= trade-duration
-        roi_list = list(filter(lambda x: x <= trade_dur, self.minimal_roi.keys()))
-        if not roi_list:
-            return None, None
-        roi_entry = max(roi_list)
-        return roi_entry, self.minimal_roi[roi_entry]
+        dynamic_roi = self.dynamic_roi
+        minimal_roi = self.minimal_roi
+        start, end = dynamic_roi['dynamic_roi_start'], dynamic_roi['dynamic_roi_end']
+
+        # if the dynamic_roi dict is defined and enabled, use it, otherwise fallback to default functionality
+        if dynamic_roi and dynamic_roi['dynamic_roi_enabled']:
```

Comment:
at first glance:
Definitely extract this from the regular roi function ...

Reply:
Okay, thanks for the feedback! Right now I have this working just fine overridden in my strategy and I wanted to contribute it to the project. I am probably in over my head though in terms of the testing and all that so I'll do my best to start with the suggestions you have made.

If I bang my head against the wall too much and don't make and significant headway in the coming days I may just end up withdrawing this PR as it is very likely above my skill level to implement it correctly, I believe strongly in the idea but I do not have a ton of confidence in my ability to implement it.

Hopefully I can get it across the finish line though!

Reply:
Okay, I did the part you asked about in terms of removing the changes from the existing function and made it into its own. I also added a bunch of error checking to ensure the required fields exist when using the various types of dynamic ROI.

If you can give what I've done so far a bit of a check before I try to dive into the unit testing bits to make sure the basis I am working from in the core parts of the code are good, I'd appreciate it. 

Once I know that I have made all of the changes, in the correct ways, in the correct places outside of the unit testing, I will tackle that part to make sure everything checks out.